### PR TITLE
added import match

### LINKED ISSUE #236: Assess and Re-create oca base_import_match

### PR DESCRIPTION
**Pull request by emjay0921**

### CODE REVIEW

**Review by reichie020212 — DISMISSED** [review]

**Comment by jeremi** [comment]
@emjay0921 can you write test for it? 

Can you confirm it could work to import Individuals based on ID number?
Can you confirm it could work to import groups based on name?

**Review by jeremi — CHANGES_REQUESTED** [review]
Add tests

**Comment by emjay0921** [comment]
> @emjay0921 can you write test for it?
> 
> Can you confirm it could work to import Individuals based on ID number? Can you confirm it could work to import groups based on name?

I don't think it will be possible to add test on importing individual based on record ID as it is not fixed (As it was needed to be added in the CSV). It is possible to add it as an xml data first but wouldn't recommend it as it will be visible in Odoo.

**Comment by emjay0921** [comment]
> Add tests

Test added @jeremi .

**Comment by jeremi** [comment]
@emjay0921 I ask ID number like the ID PASS number, or passport number...
Similarly it can be on phone number or email address.

**Comment by emjay0921** [comment]
> @emjay0921 I ask ID number like the ID PASS number, or passport number... Similarly it can be on phone number or email address.

Added a sub-field to set the right field to match if the field is one2many, many2one or many2many.

**Comment by jeremi** [comment]
Can we have tests for this last changes?

File: spp_import_match/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+{
+    "name": "OpenSPP Import Matching",
+    "category": "OpenSPP",
+    "version": "15.0.0.0.0",
+    "sequence": 1,
+    "author": "OpenSPP.org",
+    "website": "https://github.com/openspp/openspp-registry",
+    "license": "LGPL-3",
+    "development_status": "Beta",
```
Comment (jeremi) [inline spp_import_match/__manifest__.py:11]: Make it Alpha

**Review by jeremi — CHANGES_REQUESTED** [review]

File: spp_import_match/models/import_match.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+
+class SPPImportMatch(models.Model):
+    _name = "spp.import.match"
+    _description = "Import Matching"
+    _order = "sequence, name"
+
+    name = fields.Char(compute="_compute_name", store=True, index=True)
+    sequence = fields.Integer(index=True)
+    model_id = fields.Many2one(
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py:13]: I think there should be an unique constraint for model_id in spp import match, to narrow down the searching, and give us back better performance

File: spp_import_match/models/import_match.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    )
+    relation = fields.Char(related="field_id.relation")
+    sub_field_id = fields.Many2one(
+        "ir.model.fields",
+        string="Sub-Field",
+        ondelete="cascade",
+        help="Sub Fields to Match in Importing",
+    )
+    match_id = fields.Many2one("spp.import.match", string="Match", ondelete="cascade")
+    model_id = fields.Many2one(related="match_id.model_id")
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py:116]: also should be an unique constraint for match_id and field_id in spp import match fields, that will be odd if we find 2 records pointing to only 1 field, right?

File: spp_import_match/models/import_match.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+        return model
+
+    @api.model
+    @tools.ormcache("model_name", "frozenset(fields)")
+    def _usable_rules(self, model_name, fields):
+        result = self
+        available = self.search([("model_name", "=", model_name)])
+        field_to_match = []
+        for record in available:
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py:86]: so if we have 2 sql constraints as I suggested, this function will be search with limit=1, save 1 for loop

**Pull request closed without merging**